### ISSUE TITLE
NullPointerException when trying to save a virtual band that has been converted to computed band

### ISSUE THREAD

**Issue by valgur** [senbox-org/s1tbx#9]
This is probably better-suited suited for snap-engine repo but I'm posting it here since it does not allow issue reporting.

Steps to produce the bug:

1. Open a product (content does not matter);
2. Do "Create Band Math from Math Expression" and use any expression (e.g. as simple as "1");
3. Do "Convert Computed Band" from the new virtual band's context menu;
4. Save the product.

Should result in a NullPointerException.

The bug appears in both S1 Toolbox v 1.1.0 and the most recent build of the GitHub repo, but not in v 1.0.1.

I dug around with a debugger a bit and I think I found the cause of the bug, but I am not quite sure how to properly fix it.

The stack trace of the exception is
```
at com.bc.ceres.core.Assert.notNull(Assert.java:94)
at org.esa.snap.jai.VirtualBandOpImage.create(VirtualBandOpImage.java:171)
at org.esa.snap.jai.VirtualBandOpImage.create(VirtualBandOpImage.java:122)
at org.esa.snap.framework.datamodel.VirtualBandMultiLevelImage$1.createImage(VirtualBandMultiLevelImage.java:65)
at com.bc.ceres.glevel.support.AbstractMultiLevelSource.getImage(AbstractMultiLevelSource.java:65)
- locked <0x2392> (a org.esa.snap.framework.datamodel.VirtualBandMultiLevelImage$1)
at com.bc.ceres.glevel.support.DefaultMultiLevelImage.getImage(DefaultMultiLevelImage.java:66)
at com.bc.ceres.glevel.MultiLevelImage.getData(MultiLevelImage.java:64)
at org.esa.snap.framework.datamodel.Band.writeRasterDataFully(Band.java:385)
at org.esa.snap.framework.dataio.ProductIO.writeAllBands(ProductIO.java:436)
at org.esa.snap.framework.dataio.ProductIO.writeProduct(ProductIO.java:390)
...
```
The object attribute causing the NullPointerException is essentially
Product→Band→sourceImage→source→sourceImage→product == null.

This seems to be caused by ConvertComputedBandIntoBandAction.actionPerformed(), where the virutalBand's values are copied over to the newly created computedBand. This includes the sourceImage attribute.
However, in the virtualBand, the the sourceImage attribute refers back to the virtualBand itself:
VirtualBand@77306 ->
MultiLevelImage sourceImage = VirtualBandMultiLevelImage@77338 ->
MultiLevelSource source = VirtualBandMultiLevelImage@77419 ->
MultiLevelImage sourceImage = VirtualBand@77306
So the computedBand still indirectly references the virtualBand.

Following that, the virtualBand is removed from the Product's bandGroup and the computedBand added in its place. Removing the virtualBand calls
```
at org.esa.snap.framework.datamodel.ProductNode.setOwner(ProductNode.java:77)
at org.esa.snap.framework.datamodel.ProductNodeGroup.notifyRemoved(ProductNodeGroup.java:302)
at org.esa.snap.framework.datamodel.ProductNodeGroup.remove(ProductNodeGroup.java:198)
at org.esa.snap.rcp.actions.ConvertComputedBandIntoBandAction.actionPerformed(ConvertComputedBandIntoBandAction.java:145)
...
```
The relevant part being
```java
public abstract class ProductNode extends ExtensibleObject {
    public void setOwner(ProductNode owner) {
        if (owner != this.owner) {
            this.owner = owner;
            product = null;
        }
    }
}
```
Which of course causes the product == null problem.

As a quick fix I simply commented out the "product = null;" line. This solved this issue for me and did not seem to break any tests except for a few which explicitly tested for this line's existence.

As for the severity of this bug, it becomes annoying quite quickly in my typical use of the S1 Toolbox. The virtual bands cannot be saved and have to be repeatedly re-computed, causing a massive slow-down.

**Comment by marpet** [comment]
Thanks for reporting this. Have you tried the beta version of SNAP 2.0?
Does this problem still exist there?
You can get it from http://step.esa.int

**Comment by valgur** [comment]
Oh, I was not aware of a beta release. I will gladly give it a try. Are there any differences between SNAP available in the public GitHub repo and the beta release (except for the beta being slightly less recent)?

Interestingly, the beta release behaves a bit differently in this context but still appears to have this bug.
My experience was as follows:
1. Did "Create Band From Math Expression" inside an opened product (the expression itself does not seem to matter).
2. Did "Convert Computed Band". The band opens nicely and looks correct.
3. Saved the product. No message of an exception appears in the UI but the log file shows
```
SEVERE [org.openide.util.RequestProcessor]: Error in RequestProcessor org.netbeans.modules.progress.ui.RunOffEDTImpl$3
java.lang.NullPointerException: product
	at com.bc.ceres.core.Assert.notNull(Assert.java:94)
	at org.esa.snap.jai.VirtualBandOpImage.create(VirtualBandOpImage.java:171)
	at org.esa.snap.jai.VirtualBandOpImage.create(VirtualBandOpImage.java:122)
	at org.esa.snap.framework.datamodel.VirtualBandMultiLevelImage$1.createImage(VirtualBandMultiLevelImage.java:65)
	at com.bc.ceres.glevel.support.AbstractMultiLevelSource.getImage(AbstractMultiLevelSource.java:65)
	at com.bc.ceres.glevel.support.DefaultMultiLevelImage.getImage(DefaultMultiLevelImage.java:66)
	at com.bc.ceres.glevel.MultiLevelImage.getData(MultiLevelImage.java:64)
	at org.esa.snap.framework.datamodel.Band.writeRasterDataFully(Band.java:385)
	at org.esa.snap.framework.dataio.ProductIO.writeAllBands(ProductIO.java:436)
	at org.esa.snap.framework.dataio.ProductIO.writeProduct(ProductIO.java:390)
	at org.esa.snap.rcp.actions.file.WriteProductOperation.writeProduct(WriteProductOperation.java:181)
	at org.esa.snap.rcp.actions.file.WriteProductOperation.run(WriteProductOperation.java:145)
	at org.netbeans.modules.progress.ui.RunOffEDTImpl$3.run(RunOffEDTImpl.java:275)
	at org.openide.util.RequestProcessor$Task.run(RequestProcessor.java:1423)
[catch] at org.openide.util.RequestProcessor$Processor.run(RequestProcessor.java:2033)
```
4. After this, the image fails to open in the Image View as well, showing
```
java.lang.NullPointerException: product
	at com.bc.ceres.core.Assert.notNull(Assert.java:94)
	at org.esa.snap.jai.VirtualBandOpImage.create(VirtualBandOpImage.java:171)
	at org.esa.snap.jai.VirtualBandOpImage.create(VirtualBandOpImage.java:122)
	at org.esa.snap.framework.datamodel.VirtualBandMultiLevelImage$1.createImage(VirtualBandMultiLevelImage.java:65)
	at com.bc.ceres.glevel.support.AbstractMultiLevelSource.getImage(AbstractMultiLevelSource.java:65)
	at com.bc.ceres.glevel.support.DefaultMultiLevelImage.getImage(DefaultMultiLevelImage.java:66)
	at org.esa.snap.jai.ImageManager.getLevelImage(ImageManager.java:890)
	at org.esa.snap.jai.ImageManager.getSourceImage(ImageManager.java:183)
	at org.esa.snap.jai.ImageManager.createColored1BandImage(ImageManager.java:395)
	at org.esa.snap.jai.ImageManager.createColoredBandImage(ImageManager.java:358)
	at org.esa.snap.glevel.BandImageMultiLevelSource.createImage(BandImageMultiLevelSource.java:116)
	at com.bc.ceres.glevel.support.AbstractMultiLevelSource.getImage(AbstractMultiLevelSource.java:65)
	at com.bc.ceres.glevel.support.ConcurrentMultiLevelRenderer.renderImpl(ConcurrentMultiLevelRenderer.java:102)
	at com.bc.ceres.glevel.support.ConcurrentMultiLevelRenderer.renderImage(ConcurrentMultiLevelRenderer.java:92)
	at com.bc.ceres.glayer.support.ImageLayer.renderLayer(ImageLayer.java:207)
	at com.bc.ceres.glayer.Layer.render(Layer.java:408)
```
in the log file.
5. After closing and re-opening this product, the Image View works correctly but the band has only what appears to be a single tile (~500x500 px) in the top-left corner.

**Comment by marpet** [comment]
We could reproduce the error. We'll try to fix it for the final SNAP 2.0 release.
Thanks again.

**Comment by valgur** [comment]
Fixed in senbox-org/snap-desktop@d92a5785bebc15c4b831fa61e298884d600b5b89.
